### PR TITLE
feat: add RFC for v2

### PR DESCRIPTION
**Pull request by sarahdayan**
A proposal that outlines the v2 of Dinero.js, its purpose and its design.

➡ [**Read the proposal**](https://github.com/dinerojs/rfcs/blob/feat/v2/0001-dinerojs-v2.md)

➡ [Dinero.js v1 on GitHub](https://github.com/sarahdayan/dinero.js)
➡ [Dinero.js v1 docs](https://sarahdayan.github.io/dinero.js/)

### CODE REVIEW

File: 0001-dinerojs-v2.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
```
Comment (Haroenv) [inline 0001-dinerojs-v2.md:134]: is there a way to still rely on the Intl result here, or should that be done with a custom currency?

File: 0001-dinerojs-v2.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+fromFloat({ amount: 499.99 });
+
+// returns a Dinero object with `amount` equal to 15000,
+// `precision` equal to `2` and `currency` equal to `USD`
+// (using the `Currency` service under the hood)
+fromFormat("$150", "$0");
+
+// returns a Dinero object with `amount` equal to 4500
+// and precision equal to `2` (using the `Currency` service under the hood)
+DineroUSD(45);
```
Comment (Haroenv) [inline 0001-dinerojs-v2.md:187]: Is this exposed for every currency, or just because USD is the default? 

I think it would be interesting to remove the default if that makes code simpler

File: 0001-dinerojs-v2.md (hunk 1 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:134]: `toFormat` would be fully decoupled from I18n, there would be no wrapping. To use I18n, you'd do:

```js
d1
  .toRoundedUnit(2)
  .toLocaleString(d1.getLocale(), options)
```

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:187]: I think it could be interesting to have one per currency: `DineroUSD`, `DineroEUR`, `DineroIQD`, etc.

That's what MoneyPHP does, they have static methods on their `MoneyFactory` class to rapidly create objects. It's quite nice because you don't have to care about passing the currency nor thinking about exponents or converting your amount into sub-units.

File: 0001-dinerojs-v2.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+);
+
+Dinero({ amount: 500 }).convert("EUR", { rates });
+```
+
+##### Integers & BigInts
+
+Dinero.js v1 uses integers for `amount` (through the `number` primitive), and this shouldn't change, but Dinero.js v2 should also allow for `bigint` [as soon as the proposal goes to stage 4](https://tc39.es/proposal-bigint/).
+
+In the meantime, and to support environments with no `bigint` support, Dinero.js v2 should let users provide their own `Calculator`. This way, they can use whatever library they want and map the calculator functions. This behavior should be made available through a Dinero factory, so it doesn't "pollute" the API.
```
Comment (osdiab) [inline 0001-dinerojs-v2.md:83]: At the interface between `Calculator` and `dinero`, would the number ever have to be coerced back to a Javascript `number`? That would break the big int guarantees as well.

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (osdiab) [reply 0001-dinerojs-v2.md:187]: Given how light passing the currency in is, this seems not particularly valuable, but don't feel strongly about it

Comment (sarahdayan) [reply 0001-dinerojs-v2.md:187]: @osdiab It's purely utilitarian. I may or may not add it in the end, one can do it easily by importing the `Currency` object.

File: 0001-dinerojs-v2.md (hunk 3 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:83]: No, they wouldn't. You'd have the choice to use either `number`s or `bigint`s.

Dinero objects created with a `bigint` amount would be incompatible with Dinero objects created with a `number` amount, as you can't calculate both types together without coercing one first. Coercing to `number` would risk losing accuracy, while coercing to `bigint` would hurt performance.

There _could_ be a conversion of `number`s into `bigint`s though, via explicitly passing a parameter. 

```js
const d1 = Dinero({ amount: 20n })
const d2 = Dinero({ amount: 20 })

d1.add(d2) // TypeError: can't convert BigInt to number
d1.add(d2, true) // Dinero object with amount: 40n
```

What do you think?

Comment (osdiab) [reply 0001-dinerojs-v2.md:83]: Hmmm I was thinking for the example you gave or legacy support using an arbitrary precision library like “big.js” by making a calculator object for it, that if dinero sets an underlying implementation for the actual data like number, then it couldn’t use the number format of the library (in the case of big, an object with multiple fields that preserves arbitrary precision); so I think for calculator to be useful, it should also have the power to dictate the underlying number format

Comment (sarahdayan) [reply 0001-dinerojs-v2.md:83]: Yes indeed, that's why you would have access to a `createDinero` utility to which you could give your own calculator. Does that make sense?

Comment (osdiab) [reply 0001-dinerojs-v2.md:83]: Yup! Just wasn’t clear from the reply above - not just the choice between number and bigint, but also potentially a totally different numeric data structure you can specify. So long as that’s the design of the calculator/createDinero utility it remains usable

File: 0001-dinerojs-v2.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+```ts
+const FOO: Currency = {
+  name: "foo",
+  symbol: "F",
+  code: "FOO",
+  exponent: 4
+};
+```
+
+#### Helpers
```
Comment (cowwoc) [inline 0001-dinerojs-v2.md:164]: My biggest beef with v1 (which is otherwise an awesome library) is the fact that you don't use constuctors. I dislike verifying method parameters using duck typing (I want to verify the entire object, not just one function at a time). Aside from `instanceof` not working, my IDE isn't willing to auto-complete methods off Dinero (whereas other libraries work just fine). I'm pretty sure using constructors would fix this as well.

So here is my 2 cents:

1. Helper factories for returning instances are fine but they should return an `instanceof Dinero`.
2. `DineroUSD`, `DineroEUR` etc feels odd. I would expect `Dinero.USD`, `Dinero.EUR` and so on. And again they would all be instance of type `Dinero`.

And to clarify, you can still return immutable instances from factory methods.

**Review by jsardev — APPROVED** [review]

File: 0001-dinerojs-v2.md (hunk 4 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:164]: Hey @cowwoc, thanks for this thorough message! I have a few follow-up command and questions:

1. When and why do you need to verify whether an object is `instanceof Dinero`? It seems to me that using TypeScript (which Dinero v2 will bring natively, but can also be done with the current third-party typings) is a much better way to do structural typing. In the case of runtime type checking, I believe the dynamic nature of prototype delegates makes `instanceof` quite unreliable. Duck typing seems much more suitable. Could you share your use case(s) with me? I'd be interested in debating this matter over an actual piece of code 🙂
2. This is rather odd that your IDE isn't auto-completing, given that a Dinero instance is a plain object. What IDE is it? Things work well in VS Code (c.f. the following screenshot from CodeSandbox). I haven't checked with WebStorm, but any IDE or editor with a JavaScript extension should provide auto-completion when trying to access object properties. This isn't specific to Dinero.
Capture d’écran 2019-09-28 à 16 08 53
3. Adding currency helpers onto the main `Dinero` object would be highly problematic, because it would weight it down with hundreds of currency definitions. To give you an idea, there are currently 165 ISO 4217 currencies and 9 crypto currencies in Dinero v2! You likely need a couple in a regular application, so it makes no sense for you to carry around a bunch of superfluous factories that you can't even tree-shake. I intend to keep currencies apart from the core, precisely because you want to only use what you need; same goes for helpers. Does that make sense?

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: duck typing is interesting, but more challenging to deal with in certain cases, I'd personally rather these were classes. I tried to implement/or extend Dinero such that I had an interface, Payment, that isa Money, but that went poorly for me in 1.0. depending on what the 2.0 timeline is expected to be, since we have a custom currency I may have to proxy/delegate to dinero so that we can support our custom currency.

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: also, not being an expert, it might be useful to look at the upcoming? Money Api for java https://www.baeldung.com/java-money-and-currency

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (xenoterracide) [reply 0001-dinerojs-v2.md:187]: I would consider retaining the existing api (to some extent) but having, e.g `currency = 'USD'`  be backed by an "object" (object as map), where the lookup key is `USD` and it maps to the CurrencyUSD object, which is then returned or delegated to.

Comment (sarahdayan) [reply 0001-dinerojs-v2.md:187]: @xenoterracide The more I'm thinking about the helpers, the less I think I'm going to add the factories for currency-flavored Dinero objects (like `DineroUSD`, `DineroEUR`, etc.)

The main reason is that if I do that, I need to provide implementations of currency objects, which is problematic because if I make the slightest mistake in a currency declaration (or if someone disagrees), we're locked in a situation where the smallest change means breaking the API. The thing is, information about currency isn't always straightforward, and often up to interpretation. In some cases, such as Arabic currencies, I'm not even qualified to make a decision because I simply do not understand the alphabet. So it's much safer for me to provide the interface and let people implement the way they want. This way, everyone does what makes sense to them, no need to vainly accommodate everyone.

Regarding your proposition (keeping string-based `currency` to use to find definitions in a dictionary, if I got your message right), it would therefore not be possible, because it implies that Dinero.js provides implementations out of the box, and as described above, it creates issues. Also, doing something like that would imply loading all the currencies internally so they can be matched from the passed string. You definitely want to avoid that from a performance standpoint.

File: 0001-dinerojs-v2.md (hunk 4 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:164]: @xenoterracide I definitely understand that coming from a Java background, classes feel more natural. Yet since this is JavaScript, the choices I'm making are different from the kind of choices you make in a 100% object-oriented language. For instance, many JavaScript developers today are embracing a functional approach, where object-oriented patterns don't make sense at all.

Can you share with me specific use cases where `instanceof` would have helped you and where duck typing fell short? And how you wanted to extend Dinero but couldn't?

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: I come from java, before that perl (and dabbled in a dozen or so total, including PHP), both support functional programming, currently working with typescript. I've never seen FP that didn't have "objects" according to what I believe the definition of an object is.

I'm not sure I need `instanceof` what I'm doing is this, currently I'm wrapping/proxying to dinero, due to its lack of support for custom currencies, but I'd just as soon replace money with dinero.

```
export default interface Quantity {
  amount: number;
}
export default interface Money extends Quantity {
  readonly currency: string;
}
export default interface Payment<T extends PaymentMethod> extends Money {
  readonly paymentMethod: T;
}
```

but my initial attempt was to extend `Dinero.Dinero` instead of `Money`, the problem that came in was that then I was expected to implement all the things it seemed when implementing `Payment`. so instead, for the time being I've created a class that proxies to dinero (though I've thought of just using `Proxy`).

basically I was looking for an easy way to extend a Dinerjo class into a Payment with an added method and found that there didn't seem to be one.

 my model is stolen shamelessly from https://amzn.to/2ZVTXRO, which is a good read for implementing currency as well.

Comment (sarahdayan) [reply 0001-dinerojs-v2.md:164]: Dinero v2 will be fully written in TypeScript, and therefore come with a `DineroFactory` interface. Would this solve your issue?

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: I mean, the `Payment` class isn't a factory, it's a `Money` so I'm not sure... maybe? I'm using the @types/dinero.js now, the problem isn't really a lack of an interface it's the lack of ability to extend (not necessarily referring to the keyword) the actual behavior without delegation. Maybe there's a better way in typescript now, but I'm not aware of it.

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: note: if there's an interface, that's not really duck typing

but further thoughts on my current problem, currently I could make my `interface Money extend DineroOptions` and then instantiate Dinero when I need to do math, with 2.0 I could probably make it so that my `Payment` type `has`a `Money` because now we can represent our "bucks" as a custom type of Money. still wondering though if there might be a way of extending the new dinero money such that I don't have to "reimplement" everything... maybe a proxy helper... ... or does typescript support mixins/traits? maybe this is relevant? https://www.typescriptlang.org/docs/handbook/mixins.html

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (thekevinbrown) [reply 0001-dinerojs-v2.md:187]: @sarahdayan I would argue that while we can't be certain that a currency won't change over time, or that we'll get it right the first time, we do have access to semver and breaking version numbers.

In my case I'm using major currencies only, and having to specify precision everywhere is annoying and error prone. And having to reimplement this lookup in every project is silly too if you ask me. The library can know how USD works and share that knowledge for everyone.

I would vastly prefer a `currency = 'USD'` style API, and if any currencies change, we use semver and release Dinero v3, marking it as a breaking change in the release notes.

File: 0001-dinerojs-v2.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
+d2.toFormat("$0,0.00"); // €100.00
```
Comment (thekevinbrown) [inline 0001-dinerojs-v2.md:135]: Is there a way to specify whether you want the resulting string to use the EUR vs € in your example? It'd be nice to be able to choose in the format string itself whether you wanted the currency symbol or the currency code I think.

File: 0001-dinerojs-v2.md (hunk 4 above)
Comment (thekevinbrown) [reply 0001-dinerojs-v2.md:164]: @xenoterracide In Typescript interfaces just define a shape. So as long as you pass something else in that matches that shape, Typescript considers it a match and is happy, regardless of if you declared usage of that particular interface in the original function or not.

I think that's what @sarahdayan means by duck typing here. You could build a custom class and start passing instances into Dinero methods without any issue.

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: @thekevinBrown, sure, but I'm trying to do something a little cleaner than this, which I think is a little cleaner than writing delegating classes or proxies

```
  const auth = await processor.authorize({
    ...Dinero({ amount: 350, currency: 'USD' }),
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```

but if I could either extend a class to simply add 1 field, or perhaps otherwise be able to write...

```
  const auth = await processor.authorize({
    amount: 350,
    currency: 'USD',
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```

Comment (thekevinbrown) [reply 0001-dinerojs-v2.md:164]: @xenoterracide Sorry, if I'm being silly, but I'm not following what your use case is. You don't need to write a delegating class or proxy for what you're doing. Why not just do:

```
const auth = await processor.authorize({
    amount: Dinero({ amount: 350, currency: 'USD' }),
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```
And have your processor accept a Dinero instance?

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: well right now, because Dinero doesn't support custom currencies, and we have one... I mean, in the future that is definitely a possibility that I've considered, maybe it's the right one... I'm not sure, the model is prescriptive that a Payment extends Money, authorize takes a Payment, but that might be a silly prescription.

Comment (xenoterracide) [reply 0001-dinerojs-v2.md:164]: although it's a little weird to have `payment.amount.getAmount()`

**Comment by xenoterracide** [comment]
is there any target date for this release?

**Comment by JasonTheAdams** [comment]
I haven't finished leafing through this entire RFC, but one thing I wanted to throw in here just in case is that for my team's use case it's important that we're able to instantiate based on the smallest currency and retrieve the smallest currency. Like Stripe, we normalize all of our amounts into the smallest possible unit for each currency.

Also, having the precision derived from the locale would be _very_ useful.

I believe this is going to continue in v2, but I thought I'd chime in just in case it's helpful. 😄

**Comment by davefp** [comment]
I've only started looking into using Dinero recently, but I wanted to leave some feedback relevant to my use-case in the hope that it is useful.

I have a webapp with a ruby backend that uses [rubymoney](https://github.com/RubyMoney/money) for its currency/money needs. This is great for the initial page render but falls down when I want to do interactive things on the frontend.

There are two small but important features missing from V1 for me:

1. Currency symbol customization(?). A lot of my users will be using CAD, and the underlying I18n implemtation wants the currency symbol to be `CA$`, which might be technically correct but isn't commonly used, especially when it's the only currency being used by a given user.
2. Automatic precision handling. I don't want round numbers of major units to display the minor unit (cents), but I have to pick the precision up front. For example, `$42` instead of `$42.00`.

After reading the RFC it seems that these could be taken care of with the currency service. Hooray! I suppose my only feedback would be to take a look at the formatting options available through rubymoney and consider including them in V2, as they're fairly comprehensive: https://www.rubydoc.info/gems/money/Money/Formatter

File: 0001-dinerojs-v2.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+- It relies on globals to facilitate the creation of many objects with the same currency. It creates a limit to a single currency at a time and makes changing the default imperative.
+
+Helpers solve that problem by letting the user pass information in various formats, or use currency factories as shortcuts.
+
+```js
+import { fromFloat, fromFormat, DineroUSD } from "@dinero.js/helpers";
+
+// returns a Dinero object with `amount` equal to `49999`
+// and `precision` equal to `2`
+fromFloat({ amount: 499.99 });
```
Comment (nickpalmer) [inline 0001-dinerojs-v2.md:178]: Instead of doing this with a special `fromFloat` helper why not allow construction with `{ unit: 499.99 }` and do the precision and rounding for the user. Those who are coming from unit storage know that there is imprecision in how the information was stored anyway, and using unit matches `toUnit()` usage. I have monkey patched this onto our Dinero usage but I will try to open a PR for accepting unit for you to consider.

File: 0001-dinerojs-v2.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+const BigIntCalculator: Calculator = {
+  add: Big.add,
+  subtract: Big.subtract
+  // ...
+};
+```
+
+#### Currency service
+
+Currently, users must specify the `precision` of their `Dinero` objects, which is set to `2` by default. When a user creates objects with a currency which exponent is different from `2` (e.g., Iraqi dinars, Japanese yens, etc.), they must specify the `precision` by hand. One way to alleviate it is by setting a `defaultPrecision`, but this only works when one is using a single currency throughout their project.
```
Comment (nickpalmer) [inline 0001-dinerojs-v2.md:110]: Why not add a currency -> precision lookup when constructing Dinero that only specify currency? This is what I have done in our monkey patched version. I will try to open a PR for this small change for you to consider. 

If Dinero is "Currency Aware" in that it knows about various currencies, then the lookup could come from the currency service and users could register information about other currencies if they need to add a crypto-coin for instance. This would be a bigger change but would fit with the move towards Currency data managed by Dinero like locale data is handled by Moment.

File: 0001-dinerojs-v2.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+The main flaws in Dinero.js v1 are how it encapsulates certain behaviors, locking users into a specific usage, and not allowing for flexibility.
+
+Two good examples are `toFormat` and `convert`:
+
+- `toFormat` wraps around the I18n API. It creates two issues:
+  - The API doesn't output an exact desired format, but a localized currency string. This behavior is confusing, because Dinero lets users pass a mask, so they reasonably expect an exact projection.
+  - The I18n API has inconsistent implementations across different environments, which makes the API unpredictable and even more confusing.
+- `convert` wraps around a provided REST API to hide the request logic. It creates two issues:
+  - It forces users to consume a REST API, preventing them from querying rates from another source (in memory, a file) or use a different protocol (SOAP, GraphQL).
+  - It performs a new network request at each call, preventing users from caching results.
```
Comment (nickpalmer) [inline 0001-dinerojs-v2.md:63]: Even more problematic is that using a Promise as the endpoint, you don't get to know what `from` and `to` are. That means you cannot set a global endpoint which can handle pulling exchange rates for various bases from OpenExchangeRates for example, and cache results based on your preferences. You have to set the convertOptions on every call to convert based on the currencies you are converting, or you have to set it up so that you return rates for EVERY possible currency pair you might run into.

Better would be to have an `endpointFactory: (from, to) => Promise`. I will try to open a PR with this enhancement for you to consider, where passing endpointFactory, or setting a defaultEndpointFactory would override endpoint.

File: 0001-dinerojs-v2.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+import { IQD } from "@dinero.js/currencies";
+
+// `precision` is set to `2` (default)
+const d1 = Dinero({ amount: 1000, currency: "IQD" });
+
+// `precision` is automatically set to `3`
+const d2 = Dinero({ amount: 1000, currency: IQD });
+```
+
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
```
Comment (nickpalmer) [inline 0001-dinerojs-v2.md:125]: s/embark/embed/ ?

File: 0001-dinerojs-v2.md (hunk 5 above)
Comment (nickpalmer) [reply 0001-dinerojs-v2.md:135]:
```
d1.toFormat("0,0.00  USD"); // 100.00 EUR
```
What would be even better is some way to say "use the exponent from the currency after the dot" so that you could use the same format for all currencies, regardless of the exponent they use. Moustache like substitutions come to mind, with exponent being converted to the correct number of zeros. This might be a clearer way to express it instead of having to use USD symbols, or maybe not.
```
d1.toFormat("0,0.{{exponent}} {{code}}"); // 100.00 EUR
```

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (nickpalmer) [reply 0001-dinerojs-v2.md:187]: I think if you have a "CurrencyRegistry", and provide default implementations which can be added to the registry if users want to operate with that currency you are in the best of all worlds. Users are free to register new currencies, or replace currencies in the registry with their preferences, but you can provide sane defaults for many currencies which users are free to use or not as they see fit.

File: 0001-dinerojs-v2.md (hunk 5 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:135]: I've been iterating on that method, and ultimately I think the best solution is to let the user provide a render function:

```js
const price = Dinero({
  amount: 500,
  currency: EUR, // This is a `Currency` object, not a string as in v1
});

price.toFormat(({ amount, currency }) => `${currency.code} ${amount}`, {
  digits: 2,
  roundingMode: Math.round,
});

// "EUR 5"
```

The render function exposes `amount`, which is the human-readable amount which you get when invoking `toRoundedUnit`; and `currency`, the `Currency` object. The second argument are options passed to `toRoundedUnit`.

This allows for much more flexibility, and is a lot more expressive when reading the code.

If users want to keep using the I18n API, they can by writing their own adapter once:

```js
function i18n({ amount, currency }) {
  return amount.toLocaleString("en-GB", {
    style: "currency",
    currency: currency.code,
  });
}

price.toFormat(i18n, {
  digits: 2,
  roundingMode: Math.round,
});

// "€5.00"
```

Or they can do it more manually, like such:

```js
const currencySymbols = {
  EUR: "€",
  USD: "$",
  // ...
};

function manualI18n({ amount, currency }) {
  const { code } = currency;
  const symbol = currencySymbols[code] || code;

  return `${symbol}${amount.toFixed(2)}`;
}

price.toFormat(manualI18n, {
  digits: 2,
  roundingMode: Math.round,
});

// "€5.00"
```

It's a lot less confusing than the mask used in v1, and it doesn't couples the library to an API which isn't consistently implemented, nor fully supported in most environments.

File: 0001-dinerojs-v2.md (hunk 6 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:178]: This complexifies the surface API for classic Dinero object creation, and makes the code do too much. Having a dedicated factory makes the code much more expressive for the end user.

File: 0001-dinerojs-v2.md (hunk 7 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:110]: Dinero v2 will be currency-aware. When creating an object, you'll pass a `Currency` object which contains the exponent of the currency:

```js
const EUR = {
  code: "EUR",
  base: 10,
  exponent: 2,
};

const d = dinero({
  amount: 500,
  currency: EUR,
});
```

Here, the `scale` of the object (formerly known as `precision`) is inferred from the currency's exponent (2 in this case). You can still pass a custom `scale` and override the exponent if you wish to specifiy a more "precise" amount.

File: 0001-dinerojs-v2.md (hunk 8 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:63]: I'm not sure I understand the issue.

File: 0001-dinerojs-v2.md (hunk 9 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:125]: Yes!

File: 0001-dinerojs-v2.md (hunk 2 above)
Comment (sarahdayan) [reply 0001-dinerojs-v2.md:187]: So, in the end there will be a separate package called `@dinero.js/currencies` with implementation for major currencies (all ISO 4217, many crypto-currencies). You can see what it looks like on the [v2 release branch](https://github.com/dinerojs/dinero.js/tree/release/v2/packages/currencies/src).

The objects are minimal, which means it's unlikely to get them wrong. If one changes for economic reasons, then this package will bump a major.

**Comment by sarahdayan** [comment]
**Thanks to everyone who read and participated in the RFC!** [Dinero v2 alpha](https://v2.dinerojs.com/docs) has been out for a week, and now is the time to close this PR and move the conversation to [GitHub Discussions](https://github.com/dinerojs/dinero.js/discussions)!

There's no fixed date for going stable or even beta yet, this will mostly depend on user feedback. As you are among the ones most involved in the project, I'm very interested in your feedback on the new API.

Feel free to open conversations in [GitHub Discussions](https://github.com/dinerojs/dinero.js/discussions), report bugs in [issues](https://github.com/dinerojs/dinero.js/issues), and reach out for any question.

**Review by ramosrodriguezy2-a11y — APPROVED** [review]
Gracias
